wallet-api/src/Language/PlutusTx/Interval.hs

```diff
@@ -0,0 +1,149 @@
+{-# LANGUAGE FlexibleContexts     #-}
+{-# LANGUAGE DerivingStrategies   #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE DeriveGeneric        #-}
+{-# LANGUAGE TemplateHaskell      #-}
+{-# LANGUAGE DeriveAnyClass       #-}
+{-# LANGUAGE MonoLocalBinds       #-}
+-- | A type for intervals and associated functions.
+module Language.PlutusTx.Interval(
+      Slot(..)
+    , Interval(..)
+    , SlotRange
+    , interval
+    , from
+    , to
+    , singleton
+    , empty
+    , always
+    , member
+    , width
+    , contains
+    , before
+    , after
+    ) where
+
+import           Codec.Serialise.Class                    (Serialise)
+import           Data.Aeson                               (FromJSON, ToJSON)
+import           Data.Swagger.Internal.Schema             (ToSchema)
+import           GHC.Generics                             (Generic)
+import           Language.Haskell.TH
+
+import           Language.PlutusTx.Lift                   (makeLift)
+
+-- | Slot number
+newtype Slot = Slot { getSlot :: Int }
+    deriving (Eq, Ord, Show, Enum)
+    deriving stock (Generic)
+    deriving anyclass (ToSchema, FromJSON, ToJSON)
+    deriving newtype (Num, Real, Integral, Serialise)
+
+makeLift ''Slot
+
+-- | A possibly open-ended inclusive-exclusive interval of slots.
+data Interval a = Interval { ivFrom :: Maybe a, ivTo :: Maybe a }
+    deriving (Eq, Ord, Show)
+    deriving stock (Generic)
+    deriving anyclass (ToSchema, FromJSON, ToJSON, Serialise)
+
+makeLift ''Interval
+
+type SlotRange = Interval Slot
+
+{- note [Definition of Interval]
+
+The purpose of this module is to provide an interval data type that 
+can be used in on-chain and off-chain code alike. Its two main uses in the 
+'wallet-api' module are the validity range of transactions, and the ranges of
+funds and slot numbers for wallet triggers.
+
+To ensure that 'Interval' can be used in on-chain code, the functions in this 
+module do not use type classes or functions from the Haskell Prelude. The types 
+of the query functions exported from here are specialised to
+'Interval Slot'. 
+
+-}
+
+
+-- | A 'SlotRange' that covers every slot.
+always :: Q (TExp SlotRange)
+always = [|| Interval Nothing Nothing ||]
+
+-- | An open-ended 'Interval' starting
+from :: Q (TExp (a -> Interval a))
+from = [|| \s -> Interval (Just s) Nothing ||]
+
+-- | An 'Interval' from negative infinity to a value (exclusive)
```

Comment:
Sometimes you say "open-ended", sometimes you talk about "infinity" - let's be consistent.

Reply:
Right, I deleted all mentions of infinity.